src/pkcs15init/pkcs15-lib.c

```diff
@@ -1880,8 +1890,9 @@ sc_pkcs15init_store_secret_key(struct sc_pkcs15_card *p15card, struct sc_profile
 	if (check_key_compatibility(p15card, keyargs->algorithm, NULL, 0, keyargs->key.data_len * 8, 0)) {
 		/* Make sure the caller explicitly tells us to store
 		 * the key as extractable. */
-		if (!(keyargs->access_flags & SC_PKCS15_PRKEY_ACCESS_EXTRACTABLE))
-			LOG_TEST_RET(ctx, SC_ERROR_INCOMPATIBLE_KEY, "Card does not support this key.");
+		/* Commented out. I don't understand why one couldn't store a key as non extractable. -HH */
+/*		if (!(keyargs->access_flags & SC_PKCS15_PRKEY_ACCESS_EXTRACTABLE))
+			LOG_TEST_RET(ctx, SC_ERROR_INCOMPATIBLE_KEY, "Card does not support this key.");*/
```

Comment:
@fabled, could you comment?

Reply:
This part is copied from sc_pkcs15init_store_private_key(). I think the purpose is that if someone wants to store a non-supported key on the token it becomes unusable unless it can be extracted. Since wrapping support is added, it would be better to check if the key can be extracted or wrapped. If changed, same change should be done in the function where this is copied from.

Reply:
I may have misunderstood return value of check_key_compatibility, thinking zero means that the key is not compatible and nonzero would mean TRUE. But it actually returns SC_SUCCESS (0) if the card can handle the key. Would it be better to change this to 
if (check_key_compatibility(...) == SC_SUCCESS) to avoid misunderstandings?

I understand the logic better now. I have to think about this a little. One thing to consider is that a card could store a CKK_GENERIC_SECRET key, but they are not used to do crypto. So there's not point in storing such key with both CKA_SENSITIVE=CK_TRUE and CKA_EXTRACTABLE=CK_FALSE.

Reply:
After all I think it is best to keep this check, so I uncommented it. It doesn't make sense to store a key which cannot be used for crypto and is not extractable. I made the check more explicit like this: if (check_key_compatibility(...) != SC_SUCCESS). I did the same modification to other functions that call this function too.